File: android/src/main/java/com/admost/reactnative/AdMostRewardedModule.java
package com.admost.reactnative;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.WritableMap;

import java.util.HashMap;

import admost.sdk.AdMostInterstitial;
import admost.sdk.listener.AdMostAdListener;

public class AdMostRewardedModule extends ReactContextBaseJavaModule {

  private static final String TAG = AdMostRewardedModule.class.getSimpleName();
  private final ReactApplicationContext reactContext;
  private final HashMap<String, AdMostInterstitial> adMostRewardedMap; // zoneId -> AdMostRewarded

  public AdMostRewardedModule(ReactApplicationContext reactContext) {
    super(reactContext);
    this.reactContext = reactContext;
    this.adMostRewardedMap = new HashMap<>();
  }

  @NonNull
  @Override
  public String getName() {
    return "AdMostRewarded";
  }

  @ReactMethod
  public void loadAd(final String zoneId) {
    AdMostInterstitial adMostRewarded = adMostRewardedMap.get(zoneId);

    if (adMostRewarded == null) {
      AdMostInterstitial newAdMostRewarded = new AdMostInterstitial(getCurrentActivity(), zoneId, getListener(zoneId));
      adMostRewardedMap.put(zoneId, newAdMostRewarded);
      newAdMostRewarded.refreshAd(false);
      return;
    }

    adMostRewarded.refreshAd(false);
  }

  @ReactMethod
  public void destroyAd(final String zoneId) {
    AdMostInterstitial adMostRewarded = adMostRewardedMap.get(zoneId);

    if (adMostRewarded != null) {
      adMostRewarded.destroy();
      adMostRewardedMap.remove(zoneId);
    }
  }

  @ReactMethod
  public void showAd(final String zoneId, final Promise promise) {
    AdMostInterstitial adMostRewarded = adMostRewardedMap.get(zoneId);

    if (adMostRewarded != null) {
      adMostRewarded.show();
      promise.resolve(true);
    } else {
      promise.reject(Constants.ADMOST_INSTANCE_NOT_FOUND, "Couldn't find any instance in this zone, call loadAd");
    }
  }

  @ReactMethod
  public void isLoading(final String zoneId, final Promise promise) {
    AdMostInterstitial adMostRewarded = adMostRewardedMap.get(zoneId);

    if (adMostRewarded != null) {
      promise.resolve(adMostRewarded.isLoading());
    } else {
      promise.reject(Constants.ADMOST_INSTANCE_NOT_FOUND, "Couldn't find any instance in this zone, call loadAd");
    }
  }

  @ReactMethod
  public void isLoaded(final String zoneId, final Promise promise) {
    AdMostInterstitial adMostRewarded = adMostRewardedMap.get(zoneId);

    if (adMostRewarded != null) {
      promise.resolve(adMostRewarded.isLoaded());
    } else {
      promise.reject(Constants.ADMOST_INSTANCE_NOT_FOUND, "Couldn't find any instance in this zone, call loadAd");
    }
  }

  private AdMostAdListener getListener(final String zoneId) {
    return new AdMostAdListener() {
      @Override
      public void onReady(String network, int ecpm) {
        WritableMap params = Arguments.createMap();
        params.putString("network", network);
        params.putInt("ecpm", ecpm);
        params.putString("zoneId", zoneId);

        AdMostModule.sendEvent(Constants.ADMOST_REWARDED_ON_READY, params);
      }

      @Override
      public void onFail(int errorCode) {
        WritableMap params = Arguments.createMap();
        params.putInt("errorCode", errorCode);
        params.putString("zoneId", zoneId);

        AdMostModule.sendEvent(Constants.ADMOST_REWARDED_ON_FAIL, params);
      }

      @Override
      public void onDismiss(String message) {
        WritableMap params = Arguments.createMap();
        params.putString("message", message);
        params.putString("zoneId", zoneId);

        AdMostModule.sendEvent(Constants.ADMOST_REWARDED_ON_DISMISS, params);
      }

      @Override
      public void onComplete(String network) {
        WritableMap params = Arguments.createMap();
        params.putString("network", network);
        params.putString("zoneId", zoneId);

        AdMostModule.sendEvent(Constants.ADMOST_REWARDED_ON_COMPLETE, params);
      }

      @Override
      public void onShown(String network) {
        WritableMap params = Arguments.createMap();
        params.putString("network", network);
        params.putString("zoneId", zoneId);

        AdMostModule.sendEvent(Constants.ADMOST_REWARDED_ON_SHOWN, params);
      }

      @Override
      public void onClicked(String network) {
        WritableMap params = Arguments.createMap();
        params.putString("network", network);
        params.putString("zoneId", zoneId);

        AdMostModule.sendEvent(Constants.ADMOST_REWARDED_ON_CLICKED, params);
      }

      @Override
      public void onStatusChanged(int statusCode) {
        WritableMap params = Arguments.createMap();
        params.putInt("statusCode", statusCode);
        params.putString("zoneId", zoneId);

        AdMostModule.sendEvent(Constants.ADMOST_REWARDED_ON_STATUS_CHANGED, params);
      }
    };
  }

}

File: ios/AdMostRewardedModule.swift
import AMRSDK
import AppTrackingTransparency
import AdSupport

@objc(AdMostRewardedModule)
class AdMostRewardedModule: NSObject, AMRRewardedVideoDelegate {

    var adMostRewardedDict: Dictionary<String, AMRRewardedVideo> = [:] // zoneId -> AdMostRewarded

    @objc static func requiresMainQueueSetup() -> Bool {
        return true
    }

    @objc
    func loadAd(
            _ zoneId: NSString,
            resolver resolve: RCTPromiseResolveBlock,
            rejecter reject: RCTPromiseRejectBlock
    ) -> Void {
        let adMostRewarded = adMostRewardedDict[zoneId as String]

        if adMostRewarded == nil {
            let newAdMostRewarded: AMRRewardedVideo = AMRRewardedVideo(forZoneId: zoneId as String)
            newAdMostRewarded.delegate = self
            adMostRewardedDict[zoneId as String] = newAdMostRewarded

            newAdMostRewarded.load()
            return;
        }

        adMostRewarded?.load()
    }

    @objc
    func destroyAd(
            _ zoneId: NSString,
            resolver resolve: RCTPromiseResolveBlock,
            rejecter reject: RCTPromiseRejectBlock
    ) -> Void {
        let adMostRewarded = adMostRewardedDict[zoneId as String]

        if adMostRewarded != nil {
            // todo: destroy adMostAd
            adMostRewardedDict.removeValue(forKey: zoneId as String)
        }
    }

    @objc
    func showAd(
            _ zoneId: NSString,
            tag adTag: NSString,
            resolver resolve: RCTPromiseResolveBlock,
            rejecter reject: RCTPromiseRejectBlock
    ) -> Void {
        let adMostRewarded = adMostRewardedDict[zoneId as String]

        if adMostRewarded != nil {
            adMostRewarded?.show(from: UIApplication.shared.delegate?.window??.rootViewController, withTag: adTag as String)
            resolve(true)
        } else {
            reject("ADMOST_INSTANCE_NOT_FOUND", "Couldn't find any instance in this zone, call loadAd", nil);
        }
    }

    @objc
    func isLoading(
            _ zoneId: NSString,
            resolver resolve: RCTPromiseResolveBlock,
            rejecter reject: RCTPromiseRejectBlock
    ) -> Void {
        let adMostRewarded = adMostRewardedDict[zoneId as String]

        if adMostRewarded != nil {
            resolve(adMostRewarded?.isLoading)
        } else {
            reject("ADMOST_INSTANCE_NOT_FOUND", "Couldn't find any instance in this zone, call loadAd", nil);
        }
    }

    @objc
    func isLoaded(
            _ zoneId: NSString,
            resolver resolve: RCTPromiseResolveBlock,
            rejecter reject: RCTPromiseRejectBlock
    ) -> Void {
        let adMostRewarded = adMostRewardedDict[zoneId as String]

        if adMostRewarded != nil {
            resolve(adMostRewarded?.isLoaded)
        } else {
            reject("ADMOST_INSTANCE_NOT_FOUND", "Couldn't find any instance in this zone, call loadAd", nil);
        }
    }

    func didReceive(_ rewardedVideo: AMRRewardedVideo!) {
        AdMostModule.instance.sendEvent(
                eventName: "ADMOST_REWARDED_ON_READY",
                body: ["network": rewardedVideo.networkName!, "ecpm": rewardedVideo.ecpm!, "zoneId": rewardedVideo.zoneId!]
        )
    }

    func didFail(toReceive rewardedVideo: AMRRewardedVideo!, error: AMRError!) {
        AdMostModule.instance.sendEvent(
                eventName: "ADMOST_REWARDED_ON_FAIL",
                body: ["errorDescription": error.errorDescription!, "errorCode": error.errorCode, "zoneId": rewardedVideo.zoneId!]
        )
    }
    
    func didDismiss(_ rewardedVideo: AMRRewardedVideo!) {
        AdMostModule.instance.sendEvent(
                eventName: "ADMOST_REWARDED_ON_DISMISS",
                body: ["zoneId": rewardedVideo.zoneId!]
        )
    }
    
    func didComplete(_ rewardedVideo: AMRRewardedVideo!) {
        AdMostModule.instance.sendEvent(
                eventName: "ADMOST_REWARDED_ON_COMPLETE",
                body: ["network": rewardedVideo.networkName!, "zoneId": rewardedVideo.zoneId!]
        )
    }

    func didShow(_ rewardedVideo: AMRRewardedVideo!) {
        AdMostModule.instance.sendEvent(
                eventName: "ADMOST_REWARDED_ON_SHOWN",
                body: ["network": rewardedVideo.networkName!, "zoneId": rewardedVideo.zoneId!]
        )
    }

    func didClick(_ rewardedVideo: AMRRewardedVideo!) {
        AdMostModule.instance.sendEvent(
                eventName: "ADMOST_REWARDED_ON_CLICKED",
                body: ["network": rewardedVideo.networkName!, "zoneId": rewardedVideo.zoneId!]
        )
    }

    func didRewardedVideoStateChanged(_ rewardedVideo: AMRRewardedVideo!, state: AMRAdState) {
        AdMostModule.instance.sendEvent(
                eventName: "ADMOST_REWARDED_ON_STATUS_CHANGED",
                body: ["zoneId": rewardedVideo.zoneId!, "statusCode": state]
        )
    }

}
